### DISCUSSION TITLE
new feature: CodeLens support

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#299] [General]
As of v0.10.3511, we have now implemented CodeLens support.  CodeLens allows us to provide extra information (and actions) inline in the code editor.  Initially we've got this one for no-argument procedures in standard modules:

<img width="440" alt="codeLens1" src="https://user-images.githubusercontent.com/25738109/127111097-98cfe976-ff29-430c-80ef-bcdbf63c0488.png">

For those that don't like it, you can of course switch it off through the project Settings:

<img width="661" alt="codeLensSetting1" src="https://user-images.githubusercontent.com/25738109/127111133-63395fa2-594b-4880-ba0a-31c0b0aeae76.png">

And we've got this codeLens warning at the top of TWIN/BAS/CLS source files that are open in VS code but are not part of the twinBASIC project (see #290):
<img width="471" alt="codeLens2" src="https://user-images.githubusercontent.com/25738109/127111154-67392200-f2cb-4aa1-83d9-842bd96b1476.png">

**Comment by WaynePhillipsEA** [comment]
## What would you like to see the CodeLens feature used for? 

Some vague _ideas_ from me;

Call references information above each procedure (e.g. "2 references" when clicked shows you the call sites).

Hit count information during the debug session  (e.g. "hit 5 times") above each procedure

Performance information during the debug session (e.g. "average time taken: 1.4s") above each procedure

Anything we add to CodeLens will be completely optional, as I'm sure some devs would prefer keeping the code editor clean.

**Reply by mwolfe02** [reply]
Will individual features be optional?  Or will it be an all or nothing?  (There are good arguments for either approach.)

**Reply by WaynePhillipsEA** [reply]
Individual.  At the moment the 'run' codeLens option is in the project Settings, but it probably makes sense to move the option(s) into a right click context menu toggle button in the code editor.

**Reply by mwolfe02** [reply]
Call references for global and module-level variables would be nice.  Also for classes themselves.

**Comment by Kr00l** [comment]
Is it intended that the CodeLens run feature shows up on API declares that have no arguments?

**Reply by WaynePhillipsEA** [reply]
> Is it intended that the CodeLens run feature shows up on API declares that have no arguments?

No, that's fixed in v0.10.4966, thanks.
